platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java

```diff
@@ -203,38 +212,48 @@ public void run() {
 	}
 
 	public boolean onGenericMotionEvent(MotionEvent event) {
-		if ((event.getSource() & InputDevice.SOURCE_JOYSTICK) == InputDevice.SOURCE_JOYSTICK && event.getAction() == MotionEvent.ACTION_MOVE) {
-			final int device_id = findJoystickDevice(event.getDeviceId());
-
+		if (event.isFromSource(InputDevice.SOURCE_JOYSTICK) && event.getAction() == MotionEvent.ACTION_MOVE) {
 			// Check if the device exists
-			if (device_id > -1) {
-				Joystick joy = mJoysticksDevices.get(device_id);
+			final int deviceId = event.getDeviceId();
+			if (mJoystickIds.containsKey(deviceId)) {
+				final int godotJoyId = mJoystickIds.get(deviceId);
+				Joystick joy = mJoysticksDevices.get(deviceId);
 
 				for (int i = 0; i < joy.axes.size(); i++) {
-					InputDevice.MotionRange range = joy.axes.get(i);
-					final float value = (event.getAxisValue(range.getAxis()) - range.getMin()) / range.getRange() * 2.0f - 1.0f;
-					final int idx = i;
-					queueEvent(new Runnable() {
-						@Override
-						public void run() {
-							GodotLib.joyaxis(device_id, idx, value);
-						}
-					});
+					final int axis = joy.axes.get(i);
+					//joystick values are pre-normalized to -1.0 to 1.0 by Android
+					final float value = event.getAxisValue(axis); // (event.getAxisValue(range.getAxis()) - range.getMin()) / range.getRange() * 2.0f - 1.0f;
+					if (!joy.axesValues.containsKey(axis) || joy.axesValues.get(axis) != value) {
```

Comment:
Can you add a comment describing the logic below.

Reply:
ok

Reply:
```java
/**
					 * As all axes are polled for each event, only fire an axis event if the value has actually changed.
					 * Prevents flooding Godot with repeated events.
					 */
```